File: 05- May/13- Count Ways To Build Good Strings/13- Count Ways To Build Good Strings (Ahmed Gamal).js
// Author: Ahmed Gamal

// You are given two integers low and high. Return the number of strings you can build such that:
// The strings are unique and only contain the characters '0' and '1'.
// The strings contain segments of zero zeros and one ones only. (can contain multiple successive segments of the same type).
// The length of each string is between low and high, inclusive.

// This is a simple Counting DP problem. We can use dp[i] to represent the number of good strings that start at index i.
// We can then build the answer by starting at the end of the string and working backwards.
// our base case is dp[i] = 1 for low <= i <= high, since we will have one good string if we end at index i.
// The transition is dp[i] = dp[i] + dp[i + zero] + dp[i + one], since we can append a 0 or 1 to any good string that ends at index i + zero or i + one.
// We can then return dp[0] as our answer.


/**
 * @param {number} low
 * @param {number} high
 * @param {number} zero
 * @param {number} one
 * @return {number}
 */
var countGoodStrings = function(low, high, zero, one) {
    // MOD is the modulo we use to avoid overflow
    const MOD = 1000000007;

    // add two numbers and take the modulo
    let add = (a, b) => (a + b) % MOD;

    // dp[i] is the number of good strings that start at index i (initialized with some additional space to avoid index out of bounds)
    let dp = Array(high << 1 | 1).fill(0);

    // base case: dp[i] = 1 for low <= i <= high
    for(let i = low; i <= high; i++) {
        dp[i] = 1;
    }  

    // transition: dp[i] = dp[i] + dp[i + zero] + dp[i + one]
    for(let i = high; ~i; --i) {
        dp[i] = add(dp[i], add(dp[i + zero], dp[i + one]));
    }

    return dp[0];
};

File: 05- May/13- Count Ways To Build Good Strings/13- Count Ways To Build Good Strings (Ahmed Gamal).cpp
// Author: Ahmed Gamal

// for this problem we will use dp (memoization)
// the idea will be the same as the iterative solution, you can see the iterative solution for more details about the idea
// we will use dp[i] to be the number of good strings we can build from i to n (n is any number in the range [low, high])

class Solution {
public:
    int countGoodStrings(int low, int high, int zero, int one) {
        // MOD is the modulo we will use to avoid overflow
        // memo is the dp array we will use to store the number of good strings we can build from i to n
        const int MOD = 1e9 + 7;
        vector<int> memo(high + 5, -1);
        
        // add is a lambda function to add two numbers and take the modulo
        auto add = [&](int& a, int b) -> void {
            a += b;
            if(a >= MOD) {
                a -= MOD;
            }
        };
        
        // dp is a lambda function to calculate the number of good strings we can build from i to n
        auto dp = [&](auto dp, int len) -> int {
            // if we reach the end of the string, we will return 1 if the length is in the range [low, high] and 0 otherwise
            if(len >= high) {
                return len == high;
            }

            // if we already calculated the number of good strings we can build from i to n, we will return it
            int& ret = memo[len];
            if(~ret) {
                return ret;
            }

            // ret will be the number of good strings we can build from i to n
            // before we add the number of good strings we can build from i to n, we will check if the length is in the range [low, high]
            // if the length is in the range, then we have a good string, so we will add 1 to the number of good strings we can build and continue building the rest of it
            ret = len >= low;

            // we will try to add a zero and a one to the string and continue building the rest of it
            add(ret, dp(dp, len + zero));
            add(ret, dp(dp, len + one));

            return ret;
        };

        // we will start building the string from the first character
        return dp(dp, 0);
    }
};